library/src/lib/calendar/calendar-views/calendar-year-view/calendar-year-view.component.ts

```diff
@@ -0,0 +1,155 @@
+import { Component, OnInit, ViewEncapsulation, Output, Input, EventEmitter, ElementRef, AfterViewChecked } from '@angular/core';
+import { FdDate } from '../../models/fd-date';
+
+/** Component representing the YearView of the Calendar Component. */
+@Component({
+    selector: 'fd-calendar-year-view',
+    templateUrl: './calendar-year-view.component.html',
+    styleUrls: ['./calendar-year-view.component.scss'],
+    encapsulation: ViewEncapsulation.None,
+    host: {
+        '[attr.id]': 'id + "-year-view"'
+    }
+})
+export class CalendarYearViewComponent implements AfterViewChecked, OnInit {
+
+    /** @hidden */
+    private newFocusedYearId: string;
+
+    /** Parameter used in id of years used for help with focusing on the correct element during keyboard navigation. */
+    @Input()
+    id: string;
+
+    /** Function that is called when the focus would escape the element. */
+    @Input()
+    focusEscapeFunction: Function;
+
+    /** Parameter that stores the dozen of years that are currently being displayed. */
+    calendarYearList: number[];
+
+    /** Parameter holding the year that is currently selected. */
+    @Input()
+    yearSelected: number;
+
+    /** Event fired when a year is selected. */
+    @Output()
+    readonly yearClicked: EventEmitter<number> = new EventEmitter<number>();
+
+    /** Parameter storing the year of the present day. */
+    currentYear: number = FdDate.getToday().year;
+    firstYearInList: number = this.currentYear;
+
+    /** @hidden */
+    ngAfterViewChecked(): void {
+        if (this.newFocusedYearId) {
+            this.focusElement(this.newFocusedYearId);
+            this.newFocusedYearId = null;
+        }
+    }
+
+    /** @hidden */
+    ngOnInit(): void {
+        this.firstYearInList = this.yearSelected;
+        this.constructYearList();
+    }
+
+    /** @hidden */
+    constructor(private eRef: ElementRef) { }
+
+    /** @hidden */
+    private constructYearList() {
+        this.calendarYearList = [];
+        for (let x = 0; x < 12; x++) {
+            this.calendarYearList.push(this.firstYearInList + x);
+        }
+    }
+
+    /** Method for handling the keyboard navigation. */
+    onKeydownYearHandler(event, year: number, index: number) {
```

Comment:
There's gotta be a way to reuse this. There's similar logic in many components here

Reply:
Right now there are some methods and subjects in calendar service which are responsible for handling keydown events.